### PR TITLE
chore(core): add proper types to public facing apis

### PR DESCRIPTION
**Pull request by Hotell**
**This PR** 
- fixes some of public facing API's within `@nrwl/workspace` 
- adds some commonly used/needed types for consumers of this package (workspace json interface and package.json  interface)
- adds types for some missing 3rd party packages
- ~sets `esModuleInterop` for workspace package (this would need a lot of work to make it work across all packages)~
- ~sets `allowSyntheticDefaultImports` on to be able to use 3rd party packages types~
- **NOTE: strict mode is still turned off** so it cannot be checked automatically (this would require rather big amount of additional work, this PR should be a "good enough" first step)

## Current Behavior
- various public facing API's from `@nrwl/workspace` are not properly typed
- there are important interfaces missing (for workspace.json, package.json)

## Expected Behavior
- public facing API's should be type correct
- consumers writing custom plugins/schematics should get best possible DX  ( when manipulating with package.json and workspace.json, it is missing )

## Related Issue(s)

Fixes #

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -205,7 +207,7 @@
     "source-map-loader": "0.2.4",
     "source-map-support": "^0.5.12",
     "speed-measure-webpack-plugin": "1.3.1",
-    "strip-json-comments": "2.0.1",
+    "strip-json-comments": "3.0.0",
```
Comment (Hotell) [inline package.json:210]: no "dangerous" breaking changes + ships with types

File: packages/workspace/index.ts [hunk 2]
```diff
@@ -25,7 +25,11 @@ export { output } from './src/utils/output';
 export { commandsObject } from './src/command-line/nx-commands';
 export { supportedNxCommands } from './src/command-line/supported-nx-commands';
 export { readWorkspaceJson, readNxJson } from './src/core/file-utils';
-export { NxJson } from './src/core/shared-interfaces';
+export {
+  NxJson,
+  NxWorkspaceJson,
```
Comment (Hotell) [inline packages/workspace/index.ts:30]: exposing 2 new interfaces

File: packages/workspace/src/command-line/utils.ts [hunk 3]
```diff
@@ -1,5 +1,5 @@
-import * as yargsParser from 'yargs-parser';
-import * as yargs from 'yargs';
+import yargsParser from 'yargs-parser';
```
Comment (Hotell) [inline packages/workspace/src/command-line/utils.ts:1]: needed as `esModuleInterop` is used

File: packages/workspace/src/core/shared-interfaces.ts [hunk 4]
```diff
@@ -41,3 +42,42 @@ export interface Environment {
...
+  projectType: 'library' | 'application';
+
+  schematic: EmptyMap | AnyStringMap;
+  architect: Record<
+    string,
+    import('@angular-devkit/core').workspaces.TargetDefinition
+  >;
+}
+
+export interface NxWorkspaceJson<P = string> {
```
Comment (Hotell) [inline packages/workspace/src/core/shared-interfaces.ts:58]: this generic is needed for schematics rules which use `join` and friends from `angular/devkit` - they don't use primitive `string` rather branded(nominal) `Path` ( which is string in reality), this properly covers these cases so API's used for schematics have it type as `Path`

File: scripts/tsconfig.scripts.json [hunk 5]
```diff
@@ -1,6 +1,7 @@
 {
   "extends": "../tsconfig.base.json",
   "compilerOptions": {
+    "esModuleInterop": true,
```
Comment (Hotell) [inline scripts/tsconfig.scripts.json:4]: this is now needed to make docs pass

**Comment by vsavkin** [comment]
Thanks for the PR!

The PR is a bit too big
- with some changes being bug fixes (e.g., `migrate.ts`)
- some replacing "any" with well-defined types (which is good)
- some are just misc changes, like:

```
export function workspaceFileName() {
  if (fileExists(`${appRootPath}/angular.json`)) {
    return 'angular.json';
  }
  return 'workspace.json';
}
```

We should probably split it into two. Add a PR with bug fixes (fix(core):...), and a PR adding more types. Adding types is probably `cleanup(core):...`. I'd remove all misc changes.

**Comment by vsavkin** [comment]
Given that the PR does so much, I think it will be challenging to land. A lot of internal work is being done for v11, so anything that touches so many things will likely to conflict with something else. I think it's better to split the PR into multiple smaller PRs. We can chat about it more in Slack. For now though I'm going to close it. Sorry about this.

**Pull request closed without merging**